### DISCUSSION TITLE
Is there a faster way to send data from rust to webview?

### DISCUSSION THREAD

**Discussion by xuchaoqian** [tauri-apps/tauri#5511] [General]
In our use case,  we need send a large set of dynamic data(about 200,000 rows) from rust to webview repeatedly. 

We use **window.eval** right now, it takes more than 16 seconds to finish data transfer. Even worse, the UI will be blocked in the meantime.

Just clone [the repo](https://github.com/xuchaoqian/tauri-examples.git) to run the demo:
`cd tauri-examples/from-rust-to-webview/ && npm install --verbose && npm run tauri dev`

Is there a faster way to send data from rust to webview?

**Comment by FabianLars** [comment]
Sending this much data will probably always be problematic. Btw, try comparing debug vs release mode, there should be a relatively big difference.

The only built-in alternative would be custom protocols https://docs.rs/tauri/latest/tauri/struct.Builder.html#method.register_uri_scheme_protocol. Here's an example app using it for image data: https://github.com/JonasKruckenberg/pisano/ - Our new IPC will very likely build on that same system and we're trying to implement streaming support too.

Other than that there are only separate things like hosting a localhost http or ws server :/

**Reply by xuchaoqian** [reply]
Thank you so much for your solutions!


> Btw, try comparing debug vs release mode, there should be a relatively big difference.

It takes 13~14 seconds in release mode.




> The only built-in alternative would be custom protocols https://docs.rs/tauri/latest/tauri/struct.Builder.html#method.register_uri_scheme_protocol

In our case, the webview uses the JS objects instead of the binary data, such as images/video, etc. 
So, we will still need serialization-deserialization, even if use custom protocols. Is it right?
Is there a way to create JS objects in the rust backend and injects them into webview directly?

> we're trying to implement streaming support too.

Btw., very interesting to see the streaming support solution. 
Can I use custom protocols to implement it right now? 
Basically, I want to implement a websocket proxy in the rust backend.

**Reply by FabianLars** [reply]
> So, we will still need serialization-deserialization, even if use custom protocols. Is it right?

Yes and no. You still need to have a shared representation format yes, but custom protocols don't require json strings as their payload which means you can send binary data without converting it to a string first. Also they _seem to_ be slightly faster even with the same payload.

> Btw., very interesting to see the streaming support solution. Can I use custom protocols to implement it right now?

No, it needs to be [implemented in Wry](https://github.com/tauri-apps/wry/pull/667) first, unless the streaming-_like_ implementation shown in our [example](https://github.com/tauri-apps/tauri/tree/dev/examples/streaming) is what you're looking for.

**Reply by xuchaoqian** [reply]
> You still need to have a shared representation format yes

Is it possible to create a [JS ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer) in rust backend, and transfer the ownership of this ArrayBuffer to webview?

**Reply by FabianLars** [reply]
Nope

**Reply by xuchaoqian** [reply]
Thanks a lot.

**Reply by notdanilo** [reply]
> > You still need to have a shared representation format yes
> 
> Is it possible to create a [JS ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer) in rust backend, and transfer the ownership of this ArrayBuffer to webview?

That would be the most optimal way to do it. I wonder what it takes to get it done.

**Reply by domehead100** [reply]
It would require shared memory (memory mapped file) to share the buffer from the backend processes to the frontends.

Should be easy to share a buffer with the Wry frontend, but then how to share it with the webview?

On Windows, the Webview2 does seem to have an api for doing this, but as far as I know the other platform webviews do not.

**Reply by FabianLars** [reply]
> On Windows, the Webview2 does seem to have an api for doing this, but as far as I know the other platform webviews do not.

Correct :/ On top of that, i've seen multiple "native" webview2 users reporting that webview2's shared memory seems to be relatively slow somehow... (didn't try it myself so i'll take that with a grain of salt) Also, implementing a good and safe API on top of shared memory is no easy task.

Anyway, i think it's realistic to say that the new v2 IPC backend is the best we can offer for the near-ish future.

**Comment by JohnScience** [comment]
I found [Medium article "Practice Rust and TAURI: Make an Image Viewer #4"](https://medium.com/@marm.nakamura/practice-rust-and-tauri-make-an-image-viewer-4-39623547b06d) quite useful.
